src/test/java/io/cdap/plugin/gcp/gcs/sink/GCSBatchSinkCmekKeyTest.java

```diff
@@ -0,0 +1,155 @@
+import io.cdap.cdap.etl.api.validation.CauseAttributes;
+import io.cdap.cdap.etl.api.validation.ValidationFailure;
+import io.cdap.cdap.etl.mock.validation.MockFailureCollector;
+import io.cdap.plugin.gcp.common.GCPConfig;
+import io.cdap.plugin.gcp.common.GCPReferenceSinkConfig;
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.internal.util.reflection.FieldSetter;
+
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * GCSBatchSink Cmek Key integration test. This test will only be run when below property is provided:
+ * project.id -- the name of the project where staging bucket may be created or new resource needs to be created.
+ * It will default to active google project if you have google cloud client installed.
+ * bucket.path -- the path of the gcs bucket
+ * service.account.file -- the path to the service account key file
+ */
+public class GCSBatchSinkCmekKeyTest {
+  private static String serviceAccountKey;
+  private static String project;
+  private static String serviceAccountFilePath;
+  private static String bucketPath;
+
+  @BeforeClass
+  public static void setupTestClass() throws Exception {
+    // Certain properties need to be configured otherwise the whole tests will be skipped.
+
+    String messageTemplate = "%s is not configured, please refer to javadoc of this class for details.";
+
+    project = System.getProperty("project.id");
+    if (project == null) {
+      project = System.getProperty("GOOGLE_CLOUD_PROJECT");
+    }
+    if (project == null) {
+      project = System.getProperty("GCLOUD_PROJECT");
+    }
+    Assume.assumeFalse(String.format(messageTemplate, "project id"), project == null);
+    System.setProperty("GCLOUD_PROJECT", project);
+
+    bucketPath = System.getProperty("bucket.path");
+    Assume.assumeFalse(String.format(messageTemplate, "bucket path"), bucketPath == null);
+
+    serviceAccountFilePath = System.getProperty("service.account.file");
+    Assume.assumeFalse(String.format(messageTemplate, "service account key file"), serviceAccountFilePath == null);
+
+    serviceAccountKey = new String(Files.readAllBytes(Paths.get(new File(serviceAccountFilePath).getAbsolutePath())),
+                                   StandardCharsets.UTF_8);
+
+  }
+
+  private GCSBatchSink.GCSBatchSinkConfig getConfig(@Nullable String project, @Nullable String path,
+                                                    @Nullable String serviceAccountType,
+                                                    @Nullable String serviceFilePath,
+                                                    @Nullable String serviceAccountJson) throws NoSuchFieldException {
+    GCSBatchSink.GCSBatchSinkConfig gcsBatchSinkConfig = new GCSBatchSink.GCSBatchSinkConfig();
+    FieldSetter
```

Comment:
Don't set fields like this. It is brittle (renaming a field will break this test).

Create a Builder for the config and create it normally.